Remove duplicate `Line_Ending_Style` and update defaults

radeusgd opened the pull request:
[ci no changelog needed]

### Pull Request Description

Implements https://www.pivotaltracker.com/story/show/182749831

### Important Notes

<!--
- Mention important elements of the design.
- Mention any notable changes to APIs.
-->

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [ ] If GUI codebase was changed: Enso GUI was tested when built using BOTH
        `./run ide dist` and `./run ide watch`.

radeusgd committed:
Remove duplicate Line_Ending_Style, change default for writing to Unix, fix tests error display

radeusgd committed:
Update docs and tests

jdunkerley commented on distribution/lib/Standard/Test/0.0.0-dev/src/Main.enso:
```diff
@@ -733,8 +733,8 @@ run_spec ~behavior =
             Finished_With_Error err stack_trace_text ->
-                Failure ("An unexpected error was returned: " + err.to_text + '\n' + stack_trace_text)
-            _ -> Failure ("An unexpected panic was thrown: " + ex.to_text + '\n' + maybeExc.get_stack_trace_text)
+                Failure ("An unexpected error was returned: " + err.to_display_text + '\n' + stack_trace_text)
```
What is this change for?

radeusgd replied:
Before:
```
Reason: An unexpected panic was thrown: (Unsupported_Argument_Types org.enso.interpreter.runtime.data.Array@53b2e1eb)
```
After:
```
An unexpected panic was thrown: Unsupported argument types: [org.enso.base.encoding.ReportingStreamEncoder, Array, Text, Unix, Text, Text, org.enso.table.write.WriteQuoteBehavior, True]
```

Although maybe the proper fix would be to fix the `Array.to_text` to print the array nicely instead of how it does now (if that is even possible).
Still I needed to see the error content and using `to_display_text` seemed as the easiest way to achieve it.

jdunkerley reviewed (approved):
One question.

Also given `Infer` now is `Unix`, do we still want `Infer`? I believe the reader ignores us anyway.

radeusgd commented:
> Also given `Infer` now is `Unix`, do we still want `Infer`? I believe the reader ignores us anyway.

It is not the same. In append mode, `Infer` will check with whatever was in the file already and use that. If `Unix` was specified, append would check if the existing file has unix endings and if not - it would fail saying that the endings mismatched. So the semantics is different. Also, specifying a specific line ending for reading also takes effect - for example reading `\r` file in Unix mode will not find the line endings. The only case is reading Windows `\r\n` in Unix mode (possibly legacy Mac too, i didn't check) where it is still read as if you specified Windows.

radeusgd committed:
fixes

jdunkerley committed:
Cope with no line endings

jdunkerley committed:
Add tests for the comment edge cases.

radeusgd committed:
Add more tests

radeusgd committed:
Implement read last n bytes

radeusgd committed:
checkpoint

radeusgd committed:
fix

radeusgd committed:
Comment quoting

radeusgd committed:
Add more edge cases

They do not work yet :(

radeusgd committed:
checkpoint

radeusgd committed:
Fix line ending edge cases

radeusgd committed:
check a few more edge cases, just to be sure

radeusgd commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoFile.java:
```diff
@@ -51,6 +54,25 @@ public InputStream inputStream(OpenOption[] opts) throws IOException {
     return this.truffleFile.newInputStream(opts);
   }
 
+  @Builtin.Method(name = "read_last_bytes_builtin")
+  @Builtin.WrapException(from = IOException.class, to = PolyglotError.class, propagate = true)
+  public Array readLastBytes(long n) throws IOException {
+    try (SeekableByteChannel channel = this.truffleFile.newByteChannel(Set.of(StandardOpenOption.READ))) {
+      int bytesToRead = Math.toIntExact(Math.min(channel.size(), n));
+      channel.position(channel.size() - bytesToRead);
+      ByteBuffer buffer = ByteBuffer.allocate(bytesToRead);
+      while (buffer.hasRemaining()) {
+        channel.read(buffer);
+      }
+
+      Object[] result = new Long[bytesToRead];
+      for (int i = 0; i < bytesToRead; ++i) {
+        result[i] = (long) buffer.get(i);
+      }
+      return new Array(result);
+    }
+  }
```
I have to admit this feels slightly sketchy - writing a very specific function as a builtin.

I guess a more general way would be to create a method that returns the `RandomAccessFile` in a similar manner to `InputStream` and then have an Enso wrapper similar to `Input_Stream`. However this seemed like an overkill for such a simple problem and when I was first implementing this I wasn't ever sure yet if we will use this finally.

I can rewrite it to create some simple `Random_Access_File` Enso wrapper if you think it is necessary, although admittedly all this buffer work seems a bit simpler to do in Java.

jdunkerley replied:
I think as this is an internal method for now happy to leave as is - if we find a need for random access more generally we can add it.

jdunkerley commented on std-bits/table/src/main/java/org/enso/table/read/DelimitedReader.java:
```diff
@@ -180,6 +180,7 @@ private CsvParser setupCsvParser(Reader input, String commentCharacter) {
       }
 
       format.setComment(commentCharacter.charAt(0));
+      settings.setCommentCollectionEnabled(true);
```
This isn't needed anymore - I don't think you are using it anymore.

jdunkerley commented on std-bits/table/src/main/java/org/enso/table/read/DelimitedReader.java:
```diff
@@ -353,18 +354,20 @@ public int getColumnCount() {
-      ensureHeadersDetected();
-      return parser.getDetectedFormat().getLineSeparatorString();
-    }
+    ensureHeadersDetected();
```
Very minor - but if newline is not null we don't need to do this.

jdunkerley reviewed (approved):
One minor change and one tiny (ignorable) comment - but approving so you can merge once fixed.

hubertp commented on test/Table_Tests/src/Util.enso:
```diff
@@ -16,7 +17,7 @@ Column.Column.should_equal expected =
     self.to_vector.should_equal expected.to_vector
 
-normalize_lines string line_separator=System.default_line_separator newline_at_end=True =
+normalize_lines string line_separator='\n' newline_at_end=True =
```
That should probably remain as a constant somewhere?

radeusgd replied:
```suggestion
normalize_lines string line_separator=Line_Ending_Style.Unix.to_text newline_at_end=True =
```

Does that look ok?

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoFile.java:
```diff
@@ -51,6 +54,25 @@ public InputStream inputStream(OpenOption[] opts) throws IOException {
+        channel.read(buffer);
+      }
+
+      Object[] result = new Long[bytesToRead];
```
This is certainly ineffective. All the content of the file is going to be split to 8-bytes chunks and an object (of type `java.lang.Long`) allocated for each of them. If you want this to be effective, then you can implement
```
class ArrayOverBuffer implements TruffleObject {
  final ByteBuffer buffer;
}
```
and return that object.

JaroslavTulach reviewed (approved):
Conversion of last few bytes to `Array` of `Long` is technically OK and preferred from my point of view than using `asGuestValue`. I'd be more happy if our builtins operated only on interop values and not on `asGuestValue` wrapped host Java objects. The rule should be:
- if you are a built in then use interop values
- once the value gets to stdlib, feel free to use Java host classes to wrap such values

radeusgd committed:
cr

radeusgd committed:
formatting

radeusgd committed:
Return a more efficient ArrayOverBuffer

radeusgd commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/data/ArrayOverBuffer.java:
```diff
@@ -0,0 +1,44 @@
+
+import java.nio.ByteBuffer;
+
+@ExportLibrary(InteropLibrary.class)
+@Builtin(pkg = "mutable", stdlibName = "Standard.Base.Data.Array.Array_Over_Buffer")
+public class ArrayOverBuffer implements TruffleObject {
```
I implemented this as @JaroslavTulach suggested, can you please take a look (together with @hubertp) to see if I didn't do something stupid here?

I'm a bit unsure of whether I should also add `MethodDispatchLibrary` here, as `Array` has it. It seemed not strictly necessary (all tests I want do pass currently) but I'm not sure if ignoring it does not introduce some hard to understand issue?

Also, shall I add Enso shadow definitions for `ArrayOverBuffer`? Ideally I think from Enso's perspective, both of these types should be seen as the same `Array` Enso type - the particular class being a particular implementation of that. But I'm not sure if this can be achieved easily? What do you think?

I guess this touches the whole discussion of revisiting arrays and having multiple possible underlying representations for them. Although it would be good if we would not block integrating this PR by getting the arrays into fully shape as this may be a much longer endeavour, so it would be cool if we can agree on a solution that can be integrated tomorrow and possibly decide to improve it later on if necessary.

radeusgd replied:
I tested the following script:
```
from Standard.Base import all

polyglot java import bar.Foo

foreign js make_js_array = """
    return [103, 101, 102, 100];

main =
    pure_enso = [40, 30]
    from_java = Vector.Vector Foo.makeJavaList
    from_js = Vector.Vector make_js_array
    bytebuffer = (enso_project.data / "test.txt").read_last_bytes 10

    [pure_enso, from_java, from_js, bytebuffer].each vec->
        IO.println vec
        IO.println vec.length
        IO.println (vec.map (+2))
        IO.println vec.sort
        IO.println ""
```

and all seems to work as expected.

I'm not sure if we shouldn't extend method dispatch for `ArrayOverBuffer`, but currently it should just behave as any foreign array we may also get - which seems to be good enough for now. In particular, it definitely solves the usecase we use this for and it is currently only an internal function anyway.

radeusgd committed:
split types

radeusgd commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/type/Types.java:
```diff
@@ -130,6 +131,8 @@ public static String getName(Object value) {
       return ConstantsGen.MANAGED_RESOURCE;
     } else if (TypesGen.isArray(value)) {
       return ConstantsGen.ARRAY;
+    } else if (TypesGen.isArrayOverBuffer(value)) {
+      return ConstantsGen.ARRAY_OVER_BUFFER;
```
Ideally we'd have these merged but I'm not sure if it is safe to do?

OTOH, as long as we use the ArrayOverBuffer only in internal functions, it probably doesn't matter, because we don't check the type anyway, _what can go wrong?_ (_famous last words_)
Still this type may get used in more places in the future so I'd prefer to do it the _right way_, but I'm not sure which way it is? cc: @JaroslavTulach @hubertp

radeusgd replied:
Discussed with @JaroslavTulach and collapsed the two array representations to report the same Enso type.

radeusgd committed:
Do not generate a separate type for ArrayOverBuffer, try to pretend it's an Array

radeusgd committed:
fix test

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/data/ArrayOverBuffer.java:
```diff
@@ -0,0 +1,43 @@
+
+  @ExportMessage
+  long getArraySize() {
+    return buffer.remaining();
```
`remaining` assumes the current `position=0`. Maybe you want to replace it with `limit`.

radeusgd replied:
Right! I assume that the buffer is in 'read mode' and the array should contain the content which would be 'read' from the buffer.

But indeed I forgot to offset the position, so I did the following update:
```diff
Object readArrayElement(long index) {
-    return (long) buffer.get(Math.toIntExact(index));
+    return (long) buffer.get(buffer.position() + Math.toIntExact(index));
}
```

Do you think that makes sense, or instead of indexing position...position+remaining, I should switch to 0...limit?

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoFile.java:
```diff
@@ -51,6 +54,23 @@ public InputStream inputStream(OpenOption[] opts) throws IOException {
+        channel.read(buffer);
+      }
+
+      buffer.flip();
```
I see. This sets the `position=0`.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/data/ArrayOverBuffer.java:
```diff
@@ -0,0 +1,43 @@
+import java.nio.ByteBuffer;
+
+@ExportLibrary(InteropLibrary.class)
+public class ArrayOverBuffer implements TruffleObject {
```
Make the class non-subclassable. Either `final` or package private or both.

radeusgd committed:
cr

radeusgd committed:
Fix Table_Date_Spec

radeusgd committed:
Wrap detecting reader in a NewlineDetector so that we can check if we can trust the parser's detected newline

radeusgd committed:
Switch back to Unix defaults

(The override was made to simulate Windows-like setting where platform default does not match our default for testing purposes.)

radeusgd committed:
align line endings in re-enabled test
